distributed/comm/ucx.py

```diff
@@ -0,0 +1,366 @@
+
+    proto, address = addr.split("://", 1)
+    return proto, address
+
+
+def _parse_host_port(address: str, default_port=None) -> tuple:
+    """
+    Parse an endpoint address given in the form "host:port".
+
+    >>> _parse_host_port("10.33.225.160:13337")
+    ("10.33.225.160", 13337)
+    """
+    if address.startswith("ucx://"):
+        _, address = _parse_address(address)
+
+    # if default port is None we select the next port availabe
+    # ucx-py does not currently support random port assignment
+    import random
+
+    default_port = default_port or random.randint(1024, 65000)
+    return parse_host_port(address, default_port=default_port)
+
+
+def _unparse_host_port(host, port=None):
+    return unparse_host_port(host, port)
+
+
+def get_endpoint_address(endpoint):
+    # TODO: ucx-py: 18
+    pass
+
+
+# ----------------------------------------------------------------------------
+# Comm Interface
+# ----------------------------------------------------------------------------
+
+
+class UCX(Comm):
+    """Comm object using UCP.
+
+    Parameters
+    ----------
+    ep : ucp.Endpoint
+        The UCP endpoint.
+    address : str
+        The address, prefixed with `ucx://` to use.
+    deserialize : bool, default True
+        Whether to deserialize data in :meth:`distributed.protocol.loads`
+
+    Notes
+    -----
+    The read-write cycle uses the following pattern:
+
+    Each msg is serialized into a number of "data" frames. We prepend these
+    real frames with two additional frames
+
+        1. is_gpus: Boolean indicator for whether the frame should be
+           received into GPU memory. Packed in '?' format. Unpack with
+           ``<n_frames>?`` format.
+        2. frame_size : Unsigned int describing the size of frame (in bytes)
+           to receive. Packed in 'Q' format, so a length-0 frame is equivalent
+           to an unsized frame. Unpacked with ``<n_frames>Q``.
+
+    The expected read cycle is
+
+    1. Read the frame describing number of frames
+    2. Read the frame describing whether each data frame is gpu-bound
+    3. Read the frame describing whether each data frame is sized
+    4. Read all the data frames.
+    """
+
+    def __init__(
+        self, ep: ucp.Endpoint, address: str, listener_instance, deserialize=True
+    ):
+        logger.debug("UCX.__init__ %s %s", address, listener_instance)
+        self._ep = ep
+        assert address.startswith("ucx")
+        self.address = address
+        self.listener_instance = listener_instance
+        default_port = next(_PORT_COUNTER)
+        self._host, self._port = _parse_host_port(address, default_port)
+        self._local_addr = None
+        self.deserialize = deserialize
+        self.comm_flag = None
+
+        # finalizer?
+
+    @property
+    def local_address(self) -> str:
+        return self._local_addr
+
+    @property
+    def peer_address(self) -> str:
+        # XXX: This isn't quite for the server (from UCXListener).
```

Comment:
I think this is also referring to https://github.com/Akshay-Venkatesh/ucx-py/issues/61, but I may be wrong